lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -3499,10 +3500,15 @@ balanceTransaction
     (Write.PParamsInAnyRecentEra era pp, timeTranslation)
         <- liftIO $ W.readNodeTipStateForTxWrite netLayer
     withWorkerCtx ctx wid liftE liftE $ \wrk -> do
-        (utxo, wallet, _txs) <- handler $ W.readWalletUTxO wrk
-        let utxoIndex =
-                Write.constructUTxOIndex $
-                Write.fromWalletUTxO utxo
+        (walletUTxO, wallet, _txs) <- handler $ W.readWalletUTxO wrk
+        let externalUTxO = parseExternalUTxO era
+        let internalUTxO = Write.fromWalletUTxO walletUTxO
+        liftHandler (guardUTxOConsistency externalUTxO internalUTxO)
+        -- When combining the external and internal UTxO sets, we give
+        -- precedence to the external UTxO set. Here we rely on the
+        -- left-biased behaviour of the Semigroup instance for Map:
+        let combinedUTxO = externalUTxO <> internalUTxO
+        let utxoIndex = Write.constructUTxOIndex combinedUTxO
```

Comment:
This would enable balanceTx /  coin-selection selecting any of the inputs provided in the "extraUTxO"? 

If the idea is that these inputs already ought to be selected, I'd find it risky — an accidentally provided superset would be included based on randomness.

However, there's also reference inputs (and in theory collateral inputs) to be considered. If these are included in the tx, they'd need to be resolved. And just because the user provides the resolution for them, it doesn't mean they should be included as normal inputs.